### ISSUE TITLE
Do you have tips to make my model work sufficiently? 

### ISSUE THREAD

**Issue by Nick19111996** [eigenvivek/DiffPose#42]
Hi Vivek, 
Thanks again for sharing your work. I am trying to use your workflow for lateral pelvic X-rays but I face some issues. The model works 50% of the time when I use a random lateral DiffDRR as input (see GIF 1).  I trained the model with a batch size of 2 images (due to memory limitations) and used a random offset with standard deviation of 40 for translations and 0.1 for the rotations.
When I use a real lateral X-ray as input the model never finds the right position of the X-ray source (See GIF 2). We don’t know the actual position of X-ray source but in GIF 3 we can see that the synthetic X-ray looks different compared to the actual X-ray (this GIF switches between synthethic and real X-ray).   
Do you maybe have some tips to make my model work sufficiently? Would adjusting the learning rate help? Or should I use better preprocessing of the real X-ray image? Our lateral X-rays are of low quality because we use a mobile C-arm (Figure 1). For preprocessing we currently only use “img = exposure.equalize_adapthist(img/np.max(img))” Do you think further preprocessing is required? If so, do you have tips regarding the preprocessing? 

Your help is deeply appreciated. Thanks in advance.

Kind regards, 

Nick 

![image1](https://github.com/user-attachments/assets/d9ea27a7-dcde-48d8-862c-944cbcf9d85b)
GIF 1: Applying model to random generated synthethic X-ray

![image2](https://github.com/user-attachments/assets/06c61d97-108e-44e0-90ff-5c7ee3862727)
GIF 2: Applying model to real X-ray (reference X-ray source is not the actual position of the X-ray source but is just a lateral X-ray) 

![image3](https://github.com/user-attachments/assets/991153c0-4483-40ed-a8c7-4c53244bc683)
GIF 3: Synthethic and real X-ray (klik on GIF for better resolution)

<img width="805" alt="input" src="https://github.com/user-attachments/assets/fa841342-2a24-466c-8ce3-5206f0e36c59">
Figure 1: Preprocessed lateral X-ray

**Comment by eigenvivek** [comment]
Hi @Nick19111996 cool images! The first thing that jumps out is the circular crop on your fluoroscopy image... I've never tested DiffDRR with an image like that, so perhaps the issue lies there. It looks like the region outside your image is white (1). Could you try setting it to black (0)? That way, those pixels won't contribute to the loss.

**Comment by eigenvivek** [comment]
Hi @Nick19111996 closing for now, feel free to reopen if you're still facing any issues!

**Comment by Nick19111996** [comment]
Hi Vivek, 

Thank you for your suggestions.

I tried making the white pixels black, but unfortunately, this did not resolve the issue. I then created some intraoperative images without the circular crop, but that didn’t work either. I think the intraoperative X-ray still looks quite different compared to the DiffDRR image. So, in addition to histogram equalization, I also tried histogram matching (see Figure 1).

![image](https://github.com/user-attachments/assets/e0792177-b726-4e63-8ae5-29fde141b05d)
Figure 1: Histogram matching

Unfortunately, this approach also didn’t resolve the issue. I then noticed that the model is very responsive to the black pixels at the borders of the image. To confirm this, I manually adjusted a DiffDRR image by adding a black upper left corner and used it as input for the model. In Figure 2, you can see the input and output images, which show that the model tries to align with the black corner.

![Black shapes](https://github.com/user-attachments/assets/40e1da75-342b-42c5-a402-c54f5cb7e865)
Figure 2: Result of adding black shape in the input image

Because of this, I thought it might help to focus more on the center of the image rather than the borders. To test this, I applied data augmentation with black triangles in the corners, black edges, and circular/oval shapes (to simulate random gas in the intestines). See Figure 3 for some examples of the augmentation.

![Afbeelding van WhatsApp op 2024-11-06 om 09 49 59_5507511c](https://github.com/user-attachments/assets/9501e51a-abd3-44a0-a183-483bb5935357)
Figure 3: Data augmentation with black triangles, black borders and oval/circular shapes. 

This approach worked on the manually adjusted DiffDRR data (Figure 4), but the issue persists with real C-arm images taken during surgery. Using images without histogram matching resulted in slightly better loss, but overall, the model still clearly does not find the correct pose (see Figure 5).

Do you have any other suggestions that could help resolve this?

Your help is greatly appreciated.

Kind regards, 

Nick 

![Black shapes after augmentation](https://github.com/user-attachments/assets/e48d2d1e-c1e3-4be9-80b0-4f27adf96e09)
Figure 4: Result of adding black shape in the input image after training with data augmentation 

![Result with real data after augmentation](https://github.com/user-attachments/assets/d5e1dbc8-c355-4c84-9ba9-7c52cadf2f61)
Figure 5: Result of applying model (trained with data augmentation) on real X-ray image
